File: instumentalshop/backend/src/main/java/com/cz/cvut/fel/instumentalshop/controller/CustomerController.java
package com.cz.cvut.fel.instumentalshop.controller;

import com.cz.cvut.fel.instumentalshop.dto.balance.in.IncreaseBalanceRequestDto;
import com.cz.cvut.fel.instumentalshop.dto.balance.out.BalanceResponseDto;
import com.cz.cvut.fel.instumentalshop.dto.user.in.UserCreationRequestDto;
import com.cz.cvut.fel.instumentalshop.dto.user.in.UserUpdateRequestDto;
import com.cz.cvut.fel.instumentalshop.dto.user.out.UserDto;
import com.cz.cvut.fel.instumentalshop.service.CustomerService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/")
public class CustomerController {

    private final CustomerService customerService;

    @PostMapping(
            path = "customers/register",
            consumes = MediaType.MULTIPART_FORM_DATA_VALUE
    )
    @PreAuthorize("permitAll()")
    public ResponseEntity<UserDto> register(
            @Valid @ModelAttribute UserCreationRequestDto dto,
            @RequestPart(name = "avatar", required = false) MultipartFile avatar
    ) {
        if (avatar != null && !avatar.isEmpty()) {
            dto.setAvatar(avatar);
        }

        UserDto created = customerService.register(dto);
        return ResponseEntity
                .status(HttpStatus.CREATED)
                .body(created);
    }

    @PutMapping("/customers/increase-balance")
    @PreAuthorize("hasAuthority('CUSTOMER')")
    public ResponseEntity<BalanceResponseDto> increaseBalance(@Valid @RequestBody IncreaseBalanceRequestDto requestDto){
        BalanceResponseDto responseDto = customerService.increaseBalance(requestDto);
        return new ResponseEntity<>(responseDto, HttpStatus.OK);
    }

    @GetMapping("/customers/balance")
    @PreAuthorize("hasAuthority('CUSTOMER')")
    public ResponseEntity<BalanceResponseDto> getBalance() {
        BalanceResponseDto customer = customerService.getBalance();
        return new ResponseEntity<>(customer, HttpStatus.OK);
    }

    @GetMapping("/customers/{id}")
    @PreAuthorize("permitAll()")
    public ResponseEntity<UserDto> getCustomerById(@PathVariable Long id) {
        UserDto customer = customerService.getCustomerById(id);
        return new ResponseEntity<>(customer, HttpStatus.OK);
    }

    @PutMapping("/customers/update")
    @PreAuthorize("hasAuthority('CUSTOMER')")
    public ResponseEntity<UserDto> updateCustomer(@Valid @RequestBody UserUpdateRequestDto requestDto) {
        UserDto customer = customerService.updateCustomer(requestDto);
        return new ResponseEntity<>(customer, HttpStatus.OK);
    }

    @DeleteMapping("/customers/delete")
    @PreAuthorize("hasAuthority('CUSTOMER')")
    public ResponseEntity<Void> deleteCustomer() {
        customerService.deleteCustomer();
        return new ResponseEntity<>(HttpStatus.OK);
    }

}

File: instumentalshop/frontend/src/controller/CustomerController.ts
import {UserCreationRequestDto} from "../dto/UserCreationRequestDto.ts";
import {UserDto} from "../dto/UserDto.ts";
import {IncreaseBalanceRequestDto} from "../dto/IncreaseBalanceRequestDto.ts";
import {BalanceResponseDto} from "../dto/BalanceResponseDto.ts";
import {UserUpdateRequestDto} from "../dto/UserUpdateRequestDto.ts";

export class CustomerController {
    private static readonly BASE = '/api/v1/customers';

    static async register(data: UserCreationRequestDto): Promise<UserDto> {
        const fd = new FormData()
        fd.append('username', data.username)
        fd.append('email', data.email)
        fd.append('password', data.password)
        fd.append('role', (data.role as any).toUpperCase());
        if (data.avatar) fd.append('avatar', data.avatar)

        const res = await fetch('/api/v1/customers/register', {
            method: 'POST',
            body: fd,
        })
        if (!res.ok) {
            const text = await res.text()
            throw new Error(text || 'Registration failed')
        }
        return res.json()
    }

    static async increaseBalance(
        dto: IncreaseBalanceRequestDto
    ): Promise<BalanceResponseDto> {
        const res = await fetch(`${this.BASE}/increase-balance`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(dto),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(err || 'Increase balance failed');
        }
        return res.json();
    }

    static async getBalance(): Promise<BalanceResponseDto> {
        const res = await fetch(`${this.BASE}/balance`, {
            method: 'GET',
            headers: { 'Content-Type': 'application/json' },
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(err || 'Fetch balance failed');
        }
        return res.json();
    }

    static async getCustomerById(id: number): Promise<UserDto> {
        const res = await fetch(`${this.BASE}/${id}`, {
            method: 'GET',
            headers: { 'Content-Type': 'application/json' },
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(err || 'Fetch user failed');
        }
        return res.json();
    }

    static async updateCustomer(
        dto: UserUpdateRequestDto
    ): Promise<UserDto> {
        const res = await fetch(`${this.BASE}/update`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(dto),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(err || 'Update user failed');
        }
        return res.json();
    }

    static async deleteCustomer(): Promise<void> {
        const res = await fetch(`${this.BASE}/delete`, {
            method: 'DELETE',
            headers: { 'Content-Type': 'application/json' },
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(err || 'Delete user failed');
        }
    }
}